src/webserver/DOSGuard.h

```diff
@@ -20,56 +20,71 @@
 #pragma once
 
 #include <boost/asio.hpp>
+#include <boost/iterator/transform_iterator.hpp>
 #include <string>
 #include <unordered_map>
 #include <unordered_set>
 
 #include <config/Config.h>
 
-class DOSGuard
+namespace clio {
+
+/**
+ * @brief A simple denial of service guard used for rate limiting.
+ *
+ * @tparam SweepHandler Type of the sweep handler
+ */
+template <typename SweepHandler>
+class BasicDOSGuard
 {
-    boost::asio::io_context& ctx_;
-    std::mutex mtx_;  // protects ipFetchCount_
+    std::reference_wrapper<SweepHandler> sweepHandler_;
```

Comment:
Why does this need to be a reference?

Reply:
What else would you implement it as? 
I'm open for suggestions!

My current understanding:
1) it can't be unset, we need some strategy for sweeping so a pointer is not a very good option
2) it can't be an object copied/moved in because the whole point of it is to be able to change it from outside of DOSGuard itself (for testing, etc.)
3) it can be a `SweepHandler &` instead but we already had that discussion a few months ago :D

Reply:
I was thinking copied/moved. Why exactly do we need to change it outside of the DOSGuard once we construct the DOSGuard? I thought the point was to just be able to pass in different kinds of sweepers, but we can do that at construction right? Maybe I'm missing something.

Reply:
How can we then test it from a unit test without going thru a real timer in a real `io_context` on a real thread?
The whole point is to be able to pretend something happened (by faking/forcing a behaviour) to see how the DOSGuard reacts (e.g. we can fake a timer calling out to `clear` by avoiding using a timer and just call the `clear` directly from a fake sweep handler). 

Now the real question is how do we test `IntervalSweepHandler` itself. We can of course set it up on a real `io_context` and then we just wait inside the unit test until timer fires and the relevant data is cleared which we can validate by checking `isOk(IP)`. Since here we are not testing the `DOSGuard` anymore but testing only the `IntervalSweepHandler` it seems strange to use a read `DOSGuard` instead of a mock. 

Ideally in such a test we create the **real** `SweepHandler` but set it up with a mocked `DOSGuard`. Then, as the timer fires, the real `SweepHandler` will call `guardMock->clear()` inside and we can check it by just monitoring that `clear` was called. In terms of a mock it will look something like this:

```cpp
template <typename SweepHandler>
struct BasicDOSGuardMock
{
    BasicDOSGuardMock(SweepHandler& handler)
    {
        handler.setup(this); // real sweep handler is given a mocked DOSGuard
    }

   // generates a `clear` member function that we can set expectations on
    MOCK_METHOD(void, clear, (), (noexcept));
};

class IntervalSweepHandlerTest : public AsioContextTest
{
protected:
    Config cfg{json::parse(JSONData)}; // real config and real sweep handler
    IntervalSweepHandler sweepHandler{cfg, ctx}; // ctx is a real io_context, comes from AsioContextTest
    BasicDOSGuardMock<IntervalSweepHandler> guard{sweepHandler}; // mocked guard given real sweep handler
};

TEST_F(IntervalSweepHandlerTest, SweepAfterInterval)
{
    EXPECT_CALL(guard, clear()).Times(AtLeast(1)); // we expect a call to `clear`
    std::this_thread::sleep_for(std::chrono::milliseconds(1500)); // wait for the call 
} // expectation should be fulfilled here
```